[IMP] website: arrow direction not changing in rtl

risr-odoo opened the pull request:
Steps to Reproduce :
1. Install a RtL language
2. Drag and drop the steps snippet and change the language ( Here Arabic (Syria) )
3. You will notice that the direction of arrow are not getting changed.

This commit fixes the issue of arrow direction of step snippet when user switch to any RTL language by applying correct `transform` and `scale` properties.
Previously, arrows maintained LTR orientation in RTL languages. Now they properly reflect RTL flow direction.

task-4286249



---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/186866.png)](https://mergebot.odoo.com/odoo/odoo/pull/186866)

divy-odoo commented:
Hello @risr-odoo, I see you have changed the two scss property `left/right` and transform which seems right to me but i don't prefer these changes in Js file, instead it should be done in scss. Here, what i think is, 

1. What if we adapt such a generic left/right property in `options.js` which doesn't need the changes for rtl? may be initialize both with the same value in `options.js`. ( Not 100% sure )
2. Update the transform property in scss by targeting `o_rtl` class.

I would appreciate it if you could spend some more time on this.
Thanks : )

risr-odoo commented:
> Hello @risr-odoo, I see you have changed the two scss property `left/right` and transform which seems right to me but i don't prefer these changes in Js file, instead it should be done in scss. Here, what i think is,
> 
> 1. What if we adapt such a generic left/right property in `options.js` which doesn't need the changes for rtl? may be initialize both with the same value in `options.js`. ( Not 100% sure )
> 2. Update the transform property in scss by targeting `o_rtl` class.
> 
> I would appreciate it if you could spend some more time on this. Thanks : )

@divy-odoo, Yes we have changed the two scss property `left/right` in 000.js file. we need the same `left` value so that we can assign it to the `right` when language is `rtl`. It is not possible to do so from the scss file so we have done it from javascript. This `left` property is added dynamically from the options.js file. when user switch to any `rtl` language then we removed the `left` property from element and added `right`.
For `transform` property , options.js is adding the `transform` property in the inline style of the element and we have to check the current value of `transform` property so that we can modify it. It is not possible to do so  from the scss .

risr-odoo commented:
> > Hello @risr-odoo, I see you have changed the two scss property `left/right` and transform which seems right to me but i don't prefer these changes in Js file, instead it should be done in scss. Here, what i think is,
> > 
> > 1. What if we adapt such a generic left/right property in `options.js` which doesn't need the changes for rtl? may be initialize both with the same value in `options.js`. ( Not 100% sure )
> > 2. Update the transform property in scss by targeting `o_rtl` class.
> > 
> > I would appreciate it if you could spend some more time on this. Thanks : )
> 
> @divy-odoo, Yes we have changed the two scss property `left/right` in 000.js file. we need the same `left` value so that we can assign it to the `right` when language is `rtl`. It is not possible to do so from the scss file so we have done it from javascript. This `left` property is added dynamically from the options.js file. when user switch to any `rtl` language then we removed the `left` property from element and added `right`. For `transform` property , options.js is adding the `transform` property in the inline style of the element and we have to check the current value of `transform` property so that we can modify it. It is not possible to do so from the scss .

`const currentTransform = connectorEls[i].style.transform;
                connectorEls[i].style.transform = currentTransform === "scale(1, -1)"
                ? "scale(-1, -1)"
                : "scale(-1, 1)";`
This is not possible from scss to do such things which we are doing in above code from javascript. In scss we can not get the current value of style from element like above.
If we want to do above changes from scss then we will have to introduce new class in options.js file and write style for that class in scss.

msh-odoo commented on addons/website/static/src/snippets/s_process_steps/002.scss:
```diff
@@ -23,6 +23,18 @@
+            transform: unset;
+            // `o_curve` uses to indicate the up side curves arrows, as a
+            // helper to switch the direction for the `rtl` languages.
+            &.o_curve {
```
As ```o_curve``` indicates up side curves so I would prefer class name ```o_up_curve```

msh-odoo reviewed (commented):
Improve commit message:
"Install a RtL language" → "Install an RTL language."
"the direction of arrow are not getting changed" → "the direction of the arrows does not change."

qsm-odoo commented on addons/website/static/src/snippets/s_process_steps/options.js:
```diff
@@ -108,8 +108,10 @@ options.registry.StepsConnector = options.Class.extend({
             this.hCurrentStepIconHeight = stepMainElementRect.height / 2;
             this.hNextStepIconHeight = nextStepMainElementRect.height / 2;
+            const widthOffset = stepMainElementRect.width / 2;
 
-            connectorEl.style.left = `calc(50% + ${stepMainElementRect.width / 2}px + 16px)`;
+            connectorEl.style.left = `calc(50% + ${widthOffset}px + 16px)`;
+            connectorEl.style.right = `calc(50% + ${widthOffset}px + 16px)`;
```
I don't understand how it could make sense to have `right` **and* `left` properties both higher than 50%. If it does make sense, this needs code comments.

divy-odoo replied:
> I don't understand how it could make sense to have right *and left properties both higher than 50%

Here, we are calculating the starting point of the connector arrow, which should align approximately with the step icon. The calculation is as follows:
`50% of the icon container (center of the icon) + the radius of the icon + some margin for better styling.`

qsm-odoo replied:
This does not answer my comment :grimacing:

qsm-odoo replied:
@divy-odoo If the CSS is already doing the exact same thing I guess it should be removed yes... to confirm

divy-odoo replied:
@qsm-odoo I have already removed it in last push, it was doing the same thing. Thanks : )

qsm-odoo commented on addons/website/static/src/snippets/s_process_steps/options.js:
```diff
@@ -127,10 +129,11 @@ options.registry.StepsConnector = options.Class.extend({
+            connectorEl.classList.remove("o_up_curve");
             if (type === "s_process_steps_connector_curved_arrow" && i % 2 === 0) {
-                connectorEl.style.transform = stepHeightDifference ? "unset" : "scale(1, -1)";
-            } else {
-                connectorEl.style.transform = "unset";
+                if (!stepHeightDifference) {
+                    connectorEl.classList.add("o_up_curve");
+                }
```
Make those 6 lines a one-liner, using `classList.toggle`

qsm-odoo commented on addons/website/static/src/snippets/s_process_steps/002.scss:
```diff
@@ -34,12 +46,6 @@
         }
     }
 
-    &.s_process_steps_connector_curved_arrow {
-        .s_process_step:nth-child(odd) .s_process_step_connector {
-            transform: scale(1, -1);
-        }
-    }
-
```
Why did we have both this and the related JS code in the snippet option? Do you know? Can you really remove it? Won't you break existing snippets in user databases?

divy-odoo replied:
Yes, it was intentionally removed. The SCSS was introduced to handle the existing snippets in the user database for the static version, but it is no longer needed in the master branch. Previously, in this [PR](https://github.com/odoo/odoo/pull/163285/files), we had already removed the same lines from 001.scss. It seems these lines were unexpectedly copied into 002.scss during the snippet redesign, before the mentioned PR was merged.

qsm-odoo reviewed (changes requested):
Commit message:
- Max 72 characters per line
- Commit title complete the sentence "when applied, this commit will ...". In this case, "when applied, this commit will arrow direction not changing in rtl" does not mean anything. It is also not precise enough. I would do `[IMP] website: change "Steps" snippet arrow direction for RTL languages`

> This commit fixes the issue of arrow direction of step snippet when user switch
to any RTL language by applying correct `transform` and `scale`
properties.
Previously, arrows maintained LTR orientation in RTL languages. Now they
properly reflect RTL flow direction.

That is a detail but: you can remove this whole paragraph. No information in there is useful, it just makes me lose some time reading it:
- "This commit fixes the issue of arrow direction of step snippet when user switch
to any RTL language" -> this is the commit title
- "by applying correct `transform` and `scale` properties." -> this is what your small diff does, it is obvious from the context, it does not need explanation. Only explain "how" you do it, if the code does not speak for itself (e.g. a huge diff and/or complex things). Kinda like a code comment: you did not comment the code here because it was obvious -> that automatically means there is no need for explanation in the commit message.
- "Previously, arrows maintained LTR orientation in RTL languages. Now they properly reflect RTL flow direction." -> this is just repeating the title / steps to reproduce again

In general, for a bug fix, there needs to be two things:
- Steps to reproduce
- Why it was broken (in this case, it never worked so no need for anything else than steps to reproduce, if the commit title is clear enough)

divy-odoo commented:
Hello @qsm-odoo, I apologize for the mistake in the commit message. I have made the necessary changes. Please let me know if anything is still missing.

thanks : )

divy-odoo commented:
Hello @qsm-odoo, Is there anything missing in this PR we are waiting for? Just a reminder : )
